File: airbyte-keycloak-setup/src/main/java/io/airbyte/keycloak/setup/ClientScopeConfigurator.java
/*
 * Copyright (c) 2020-2024 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.keycloak.setup;

import jakarta.inject.Singleton;
import jakarta.ws.rs.core.Response;
import java.util.Arrays;
import java.util.Map;
import java.util.Optional;
import lombok.extern.slf4j.Slf4j;
import org.keycloak.admin.client.resource.RealmResource;
import org.keycloak.representations.idm.ClientScopeRepresentation;
import org.keycloak.representations.idm.ProtocolMapperRepresentation;

/**
 * This class is responsible for the OpenID Client Scope and Mappers. This so that the API Keys will
 * work with our JWTs.
 */
@Singleton
@Slf4j
public class ClientScopeConfigurator {

  public static final int HTTP_STATUS_CREATED = 201;

  /**
   * This method creates the client scope and mappers for the Airbyte realm.
   *
   * @param keycloakRealm the realm to create the client scope in
   */
  public void configureClientScope(final RealmResource keycloakRealm) {
    final ClientScopeRepresentation clientScopeRepresentation = createClientScopeRepresentation();

    final Optional<ClientScopeRepresentation> existingClientScope = keycloakRealm.clientScopes().findAll().stream()
        .filter(scope -> scope.getName().equals(clientScopeRepresentation.getName()))
        .findFirst();

    if (existingClientScope.isPresent()) {
      clientScopeRepresentation.setId(existingClientScope.get().getId());
      keycloakRealm.clientScopes().get(existingClientScope.get().getId()).update(clientScopeRepresentation);
    } else {
      try (final Response response = keycloakRealm.clientScopes().create(clientScopeRepresentation)) {
        if (response.getStatus() == HTTP_STATUS_CREATED) {
          log.info("ClientScope {} created successfully.", clientScopeRepresentation.getName());
        } else {
          final String errorMessage = String.format("Failed to create Client Scope.\nReason: %s\n Response: %s",
              response.getStatusInfo().getReasonPhrase(), response.readEntity(String.class));
          log.error(errorMessage);
          throw new RuntimeException(errorMessage);
        }
      }
    }
  }

  /**
   * This method creates the client scope representation.
   *
   * @return the client scope representation
   */
  private ClientScopeRepresentation createClientScopeRepresentation() {
    final ClientScopeRepresentation clientScopeRepresentation = new ClientScopeRepresentation();
    clientScopeRepresentation.setName("openid");
    clientScopeRepresentation.setProtocol("openid-connect");
    clientScopeRepresentation.setProtocolMappers(Arrays.asList(buildUserIdMapper(), buildSubMapper()));
    return clientScopeRepresentation;
  }

  /**
   * This method creates the user id mapper.
   *
   * @return the user id mapper
   */
  private ProtocolMapperRepresentation buildUserIdMapper() {
    final ProtocolMapperRepresentation userIdMapper = new ProtocolMapperRepresentation();
    userIdMapper.setName("user_id");
    userIdMapper.setProtocol("openid-connect");
    userIdMapper.setProtocolMapper("oidc-usermodel-attribute-mapper");
    userIdMapper.setConfig(
        Map.of(
            "user.attribute", "user_id",
            "claim.name", "user_id",
            "jsonType.label", "",
            "id.token.claim", "true",
            "access.token.claim", "true",
            "userinfo.token.claim", "true",
            "multivalued", "false",
            "aggregate.attrs", "false"));
    return userIdMapper;
  }

  /**
   * This method creates the sub mapper.
   *
   * @return the sub mapper
   */
  private ProtocolMapperRepresentation buildSubMapper() {
    final ProtocolMapperRepresentation subMapper = new ProtocolMapperRepresentation();
    subMapper.setName("sub");
    subMapper.setProtocol("openid-connect");
    subMapper.setProtocolMapper("oidc-usermodel-attribute-mapper");
    subMapper.setConfig(
        Map.of(
            "user.attribute", "user_id",
            "claim.name", "sub",
            "jsonType.label", "",
            "id.token.claim", "true",
            "access.token.claim", "true",
            "userinfo.token.claim", "true",
            "multivalued", "false",
            "aggregate.attrs", "false"));
    return subMapper;
  }

}

File: airbyte-commons-auth/src/main/kotlin/io/airbyte/commons/auth/keycloak/ClientScopeConfigurator.kt
/*
 * Copyright (c) 2020-2026 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.auth.keycloak

import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.inject.Singleton
import jakarta.ws.rs.core.Response
import org.keycloak.admin.client.resource.RealmResource
import org.keycloak.representations.idm.ClientScopeRepresentation
import org.keycloak.representations.idm.ProtocolMapperRepresentation

private val logger = KotlinLogging.logger {}

/**
 * This class is responsible for the OpenID Client Scope and Mappers. This so that the API Keys will
 * work with our JWTs.
 */
@Singleton
class ClientScopeConfigurator {
  companion object {
    const val HTTP_STATUS_CREATED: Int = 201
  }

  /**
   * This method creates the client scope and mappers for the Airbyte realm.
   *
   * @param keycloakRealm the realm to create the client scope in
   */
  fun configureClientScope(keycloakRealm: RealmResource) {
    val clientScopeRepresentation = createClientScopeRepresentation()
    val existingClientScope = keycloakRealm.clientScopes().findAll().find { it.name == clientScopeRepresentation.name }

    existingClientScope?.let {
      updateExistingClientScope(it, clientScopeRepresentation, keycloakRealm)
    } ?: createNewClientScope(clientScopeRepresentation, keycloakRealm)
  }

  private fun updateExistingClientScope(
    scope: ClientScopeRepresentation,
    clientScopeRepresentation: ClientScopeRepresentation,
    keycloakRealm: RealmResource,
  ) {
    clientScopeRepresentation.id = scope.id
    keycloakRealm.clientScopes()[scope.id].update(clientScopeRepresentation)
  }

  private fun createNewClientScope(
    clientScopeRepresentation: ClientScopeRepresentation,
    keycloakRealm: RealmResource,
  ) {
    keycloakRealm.clientScopes().create(clientScopeRepresentation).use { response ->
      when (response.status) {
        HTTP_STATUS_CREATED -> logger.info { "ClientScope ${clientScopeRepresentation.name} created successfully." }
        else -> handleCreationError(response)
      }
    }
  }

  private fun handleCreationError(response: Response) {
    val errorMessage = "Failed to create Client Scope.\nReason: ${response.statusInfo.reasonPhrase}\nResponse: ${
      response.readEntity(
        String::class.java,
      )
    }"
    logger.error { errorMessage }
    throw RuntimeException(errorMessage)
  }

  /**
   * This method creates the client scope representation.
   *
   * @return the client scope representation
   */
  private fun createClientScopeRepresentation(): ClientScopeRepresentation {
    val clientScopeRepresentation = ClientScopeRepresentation()
    clientScopeRepresentation.name = "openid"
    clientScopeRepresentation.protocol = "openid-connect"
    clientScopeRepresentation.protocolMappers = listOf(buildUserIdMapper(), buildSubMapper())
    return clientScopeRepresentation
  }

  /**
   * This method creates the user id mapper.
   *
   * @return the user id mapper
   */
  private fun buildUserIdMapper(): ProtocolMapperRepresentation {
    val userIdMapper = ProtocolMapperRepresentation()
    userIdMapper.name = "user_id"
    userIdMapper.protocol = "openid-connect"
    userIdMapper.protocolMapper = "oidc-usermodel-attribute-mapper"
    userIdMapper.config =
      mapOf(
        "user.attribute" to "user_id",
        "claim.name" to "user_id",
        "jsonType.label" to "",
        "id.token.claim" to "true",
        "access.token.claim" to "true",
        "userinfo.token.claim" to "true",
        "multivalued" to "false",
        "aggregate.attrs" to "false",
      )
    return userIdMapper
  }

  /**
   * This method creates the sub mapper.
   *
   * @return the sub mapper
   */
  private fun buildSubMapper(): ProtocolMapperRepresentation {
    val subMapper = ProtocolMapperRepresentation()
    subMapper.name = "sub"
    subMapper.protocol = "openid-connect"
    subMapper.protocolMapper = "oidc-usermodel-attribute-mapper"
    subMapper.config =
      mapOf(
        "user.attribute" to "user_id",
        "claim.name" to "sub",
        "jsonType.label" to "",
        "id.token.claim" to
          "true",
        "access.token.claim" to
          "true",
        "userinfo.token.claim" to
          "true",
        "multivalued" to
          "false",
        "aggregate.attrs" to
          "false",
      )
    return subMapper
  }
}
